File: wow-core/src/test/kotlin/me/ahoo/wow/event/DomainEventStreamTest.kt
/*
 * Copyright [2021-present] [ahoo wang <ahoowang@qq.com> (https://github.com/Ahoo-Wang)].
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.ahoo.wow.event

import me.ahoo.test.asserts.assert
import me.ahoo.wow.messaging.DefaultHeader
import me.ahoo.wow.modeling.aggregateId
import me.ahoo.wow.modeling.toNamedAggregate
import me.ahoo.wow.serialization.event.JsonDomainEvent
import me.ahoo.wow.serialization.toJsonNode
import org.junit.jupiter.api.Test
import tools.jackson.databind.node.ObjectNode

class DomainEventStreamTest {

    private val aggregateId = FIXTURE_NAMED_AGGREGATE.toNamedAggregate().aggregateId("stream-aggregate")

    @Test
    fun `simple stream derives stream fields from first event`() {
        val event = FixtureNamedEvent("created").toDomainEvent(
            id = "event-1",
            aggregateId = aggregateId,
            commandId = "command-1",
            ownerId = "owner-1",
            spaceId = "space-1",
            version = 3,
            createTime = 1000,
        )
        val stream = SimpleDomainEventStream(
            id = "stream-1",
            requestId = "request-1",
            header = DefaultHeader.empty().with("stream", "source"),
            body = listOf(event),
        )

        stream.aggregateId.assert().isEqualTo(aggregateId)
        stream.contextName.assert().isEqualTo(aggregateId.contextName)
        stream.aggregateName.assert().isEqualTo(aggregateId.aggregateName)
        stream.ownerId.assert().isEqualTo("owner-1")
        stream.spaceId.assert().isEqualTo("space-1")
        stream.commandId.assert().isEqualTo("command-1")
        stream.version.assert().isEqualTo(3)
        stream.createTime.assert().isEqualTo(1000)
        stream.size.assert().isEqualTo(1)
        stream.first().assert().isSameAs(event)
    }

    @Test
    fun `copy materializes independent headers for simple domain events`() {
        val eventHeader = DefaultHeader.empty().with("event", "original")
        val event = FixtureNamedEvent().toDomainEvent(
            aggregateId = aggregateId,
            commandId = "command-1",
            header = eventHeader,
        )
        val stream = SimpleDomainEventStream(
            requestId = "request-1",
            header = DefaultHeader.empty().with("stream", "original"),
            body = listOf(event),
        )

        val copied = stream.copy()
        copied.header["stream"] = "copied"
        copied.first().header["event"] = "copied"

        stream.header["stream"].assert().isEqualTo("original")
        stream.first().header["event"].assert().isEqualTo("original")
    }

    @Test
    fun `copy materializes independent headers for json domain events`() {
        val sourceEvent = FixtureNamedEvent().toDomainEvent(
            aggregateId = aggregateId,
            commandId = "command-1",
        )
        val jsonEvent = JsonDomainEvent(
            id = sourceEvent.id,
            header = DefaultHeader.empty().with("event", "json-original"),
            bodyType = FixtureNamedEvent::class.java.name,
            body = """{"value":"json"}""".toJsonNode<ObjectNode>(),
            aggregateId = sourceEvent.aggregateId,
            ownerId = sourceEvent.ownerId,
            spaceId = sourceEvent.spaceId,
            version = sourceEvent.version,
            sequence = sourceEvent.sequence,
            revision = sourceEvent.revision,
            commandId = sourceEvent.commandId,
            name = sourceEvent.name,
            isLast = sourceEvent.isLast,
            createTime = sourceEvent.createTime,
        )
        val stream = SimpleDomainEventStream(
            requestId = "request-1",
            body = listOf(jsonEvent),
        )

        val copiedEvent = stream.copy().first() as JsonDomainEvent
        copiedEvent.header["event"] = "json-copied"

        jsonEvent.header["event"].assert().isEqualTo("json-original")
    }

    @Test
    fun `ignoreSourcing only ignores initial error events marked as ignore sourcing`() {
        val normalInitial = SimpleDomainEventStream(
            requestId = "request-1",
            body = listOf(
                FixtureNamedEvent().toDomainEvent(
                    aggregateId = aggregateId,
                    commandId = "command-1",
                    version = 1,
                )
            ),
        )
        val ignoredInitial = SimpleDomainEventStream(
            requestId = "request-2",
            body = listOf(
                FixtureIgnoredErrorEvent("failed", "failed event").toDomainEvent(
                    aggregateId = aggregateId,
                    commandId = "command-2",
                    version = 1,
                )
            ),
        )
        val ignoredButNotInitial = SimpleDomainEventStream(
            requestId = "request-3",
            body = listOf(
                FixtureIgnoredErrorEvent("failed", "failed event").toDomainEvent(
                    aggregateId = aggregateId,
                    commandId = "command-3",
                    version = 2,
                )
            ),
        )

        normalInitial.ignoreSourcing().assert().isFalse()
        ignoredInitial.ignoreSourcing().assert().isTrue()
        ignoredButNotInitial.ignoreSourcing().assert().isFalse()
    }
}

File: typescript/wow-client/test/query/event/domainEventStream.test.ts
/*
 * Copyright [2021-present] [ahoo wang <ahoowang@qq.com> (https://github.com/Ahoo-Wang)].
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { describe, expect, it } from 'vitest';
import {
  DomainEvent,
  DomainEventStream,
  DomainEventStreamMetadataFields,
} from '../../../src';

interface TestEventBody {
  message: string;
  count: number;
}

describe('DomainEvent', () => {
  it('should create a domain event with required properties', () => {
    const event: DomainEvent<TestEventBody> = {
      id: 'event-id-1',
      name: 'TestEvent',
      body: {
        message: 'test message',
        count: 42,
      },
      bodyType: 'TestEventBody',
      revision: '1.0',
    };

    expect(event).toBeDefined();
    expect(event.id).toBe('event-id-1');
    expect(event.name).toBe('TestEvent');
    expect(event.body.message).toBe('test message');
    expect(event.body.count).toBe(42);
    expect(event.bodyType).toBe('TestEventBody');
    expect(event.revision).toBe('1.0');
  });
});

describe('DomainEventStream', () => {
  it('should create a domain event stream with required properties', () => {
    const events: DomainEvent<TestEventBody>[] = [
      {
        id: 'event-id-1',
        name: 'TestEvent',
        body: {
          message: 'test message 1',
          count: 1,
        },
        bodyType: 'TestEventBody',
        revision: '1.0',
      },
      {
        id: 'event-id-2',
        name: 'TestEvent',
        body: {
          message: 'test message 2',
          count: 2,
        },
        bodyType: 'TestEventBody',
        revision: '1.0',
      },
    ];

    const stream: DomainEventStream = {
      id: 'stream-id-1',
      contextName: 'context-name-1',
      aggregateName: 'aggregate-name-1',
      aggregateId: 'aggregate-id-1',
      tenantId: 'tenant-id-1',
      ownerId: 'owner-id-1',
      commandId: 'command-id-1',
      createTime: Date.now(),
      requestId: 'request-id-1',
      version: 1,
      body: events,
      headers: {
        remote_ip: '127.0.0.1',
      },
    };

    expect(stream).toBeDefined();
    expect(stream.id).toBe('stream-id-1');
    expect(stream.contextName).toBe('context-name-1');
    expect(stream.aggregateName).toBe('aggregate-name-1');
    expect(stream.aggregateId).toBe('aggregate-id-1');
    expect(stream.tenantId).toBe('tenant-id-1');
    expect(stream.ownerId).toBe('owner-id-1');
    expect(stream.commandId).toBe('command-id-1');
    expect(stream.requestId).toBe('request-id-1');
    expect(stream.version).toBe(1);
    expect(stream.body).toHaveLength(2);
    expect(stream.headers.remote_ip).toBe('127.0.0.1');
  });
});

describe('DomainEventStreamMetadataFields', () => {
  it('should have correct field values', () => {
    expect(DomainEventStreamMetadataFields.HEADERS).toBe('headers');
    expect(DomainEventStreamMetadataFields.COMMAND_OPERATOR).toBe('headers.command_operator');
    expect(DomainEventStreamMetadataFields.AGGREGATE_ID).toBe('aggregateId');
    expect(DomainEventStreamMetadataFields.TENANT_ID).toBe('tenantId');
    expect(DomainEventStreamMetadataFields.OWNER_ID).toBe('ownerId');
    expect(DomainEventStreamMetadataFields.COMMAND_ID).toBe('commandId');
    expect(DomainEventStreamMetadataFields.REQUEST_ID).toBe('requestId');
    expect(DomainEventStreamMetadataFields.VERSION).toBe('version');
    expect(DomainEventStreamMetadataFields.BODY).toBe('body');
    expect(DomainEventStreamMetadataFields.BODY_ID).toBe('body.id');
    expect(DomainEventStreamMetadataFields.BODY_NAME).toBe('body.name');
    expect(DomainEventStreamMetadataFields.BODY_TYPE).toBe('body.bodyType');
    expect(DomainEventStreamMetadataFields.BODY_REVISION).toBe('body.revision');
    expect(DomainEventStreamMetadataFields.BODY_BODY).toBe('body.body');
    expect(DomainEventStreamMetadataFields.CREATE_TIME).toBe('createTime');
  });
});